ocaml/typing/solver.ml

```diff
@@ -635,6 +635,167 @@ module Solver_mono (C : Lattices_mono) = struct
+  If a [v, f] in [u.vupper] has bounds that are equal: [u] <= [f v.upper] *)
+  let tighten_bound
+      : type a. log:_ -> a C.obj -> a var -> unit =
+    fun ~log dst u ->
+      List.iter
+        (fun (Amorphvar(v, f)) ->
+          if v.lower = v.upper then begin
+            let a' = C.apply dst f v.lower in
+            let r = submode_cv ~log dst a' u in
+            assert (r |> Result.is_ok)
+          end)
+        u.vlower;
+      List.iter
+        (fun (Amorphvar(v, f)) ->
+          if v.lower = v.upper then begin
+            let a' = C.apply dst f v.lower in
+            let r = submode_vc ~log dst u a' in
+            assert (r |> Result.is_ok)
+          end)
+        u.vupper;
+      if u.lower = u.upper then begin
+        set_vlower ~log u [];
+        set_vupper ~log u []
+      end
+
+  (* Tightens the bounds of all reachable variables above [generic_level], in reverse
+  order *)
+  let tighten_bounds
+      : type a. log:_ -> a C.obj -> generic_level:int -> a var -> unit =
+    fun ~log dst ~generic_level u ->
+      let rec loop : type a. traversed:_ -> a C.obj -> a var -> unit =
+        fun ~traversed dst u ->
+          if Hashtbl.mem traversed u.id || u.level < generic_level then ()
+          else begin
+            Hashtbl.add traversed u.id ();
+            let do_loop (Amorphvar(v, f)) =
+              let src = C.src dst f in
+              loop ~traversed src v
+            in
+            List.iter do_loop u.vupper;
+            List.iter do_loop u.vlower;
+            tighten_bound ~log dst u
+          end
+      in
+      let traversed = Hashtbl.create 17 in
+      loop ~traversed dst u
+
+  (* Updates reachable variable whose level is above generic_level to
+  [generic_level]. Note that the set of reachable variables might change during an
+  iteration. Variables that remain at [generic_level + n] may be targets for garbage
+  collection *)
+  let rec update_to_generic
+      : type a. log:_ -> a C.obj -> generic_level:int -> a var -> unit =
+    fun ~log dst ~generic_level u ->
+      if u.level <= generic_level then ()
+      else begin
+        update_level_v ~log dst generic_level u;
+        let do_updgen (Amorphvar (v, f)) =
+          let src = C.src dst f in
+          update_to_generic ~log src ~generic_level v
+        in
+        List.iter do_updgen u.vupper;
+        List.iter do_updgen u.vlower
+      end
+
+  (* Updates reachable variable whose level is above generic_level AND whose bounds are
+  equal to [generic_level]. Note that the set of reachable variables might change
+  during an iteration. Variables that remain at [generic_level + n] may be targets for
+  garbage collection *)
+  let rec update_to_generic_equal_bounds
+      : type a. log:_ -> a C.obj -> generic_level:int -> a var -> unit =
+    fun ~log dst ~generic_level u ->
+      if u.level <= generic_level then ()
+      else begin
+        if u.lower = u.upper then
```

Comment:
If the bounds are not tight we want to move these variables back to `current_level`. The idea is that the only nodes whose level is `generic_level` are those with tight bounds, since they are equivalent to "structure". Variables whose values are not yet completely determined should not be considered generic as that would amount to introducing polymorphism.